// @po-ui/ng-components/fesm2022/po-ui-ng-components-po-chart-modal-table.component-C8jlPZLV.mjs
import * as i0 from '@angular/core';
import { Component, ViewChild, Input } from '@angular/core';
import { PoModalModule, PoModalComponent, PoTableModule, PoTableComponent } from './po-ui-ng-components.mjs';

class PoChartModalTableComponent {
    modalComponent;
    title;
    itemsTable;
    columnsTable;
    actionModal;
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.0.7", ngImport: i0, type: PoChartModalTableComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "20.0.7", type: PoChartModalTableComponent, isStandalone: true, selector: "po-chart-modal-table", inputs: { title: "title", itemsTable: "itemsTable", columnsTable: "columnsTable", actionModal: "actionModal" }, viewQueries: [{ propertyName: "modalComponent", first: true, predicate: ["modalComponent"], descendants: true, static: true }], ngImport: i0, template: `
    <po-modal #modalComponent [p-click-out]="true" [p-title]="title" [p-primary-action]="actionModal">
      <po-table [p-hide-columns-manager]="true" [p-items]="itemsTable" [p-columns]="columnsTable"></po-table>
    </po-modal>
  `, isInline: true, dependencies: [{ kind: "ngmodule", type: PoModalModule }, { kind: "component", type: PoModalComponent, selector: "po-modal" }, { kind: "ngmodule", type: PoTableModule }, { kind: "component", type: PoTableComponent, selector: "po-table" }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.0.7", ngImport: i0, type: PoChartModalTableComponent, decorators: [{
            type: Component,
            args: [{
                    standalone: true,
                    imports: [PoModalModule, PoTableModule],
                    selector: 'po-chart-modal-table',
                    template: `
    <po-modal #modalComponent [p-click-out]="true" [p-title]="title" [p-primary-action]="actionModal">
      <po-table [p-hide-columns-manager]="true" [p-items]="itemsTable" [p-columns]="columnsTable"></po-table>
    </po-modal>
  `
                }]
        }], propDecorators: { modalComponent: [{
                type: ViewChild,
                args: ['modalComponent', { static: true }]
            }], title: [{
                type: Input
            }], itemsTable: [{
                type: Input
            }], columnsTable: [{
                type: Input
            }], actionModal: [{
                type: Input
            }] } });

export { PoChartModalTableComponent };
